qcfractal/interface/collections/dataset.py

```diff
@@ -530,24 +553,27 @@ def add_rxn(self, name, stoichiometry, reaction_results=None, attributes=None, o
         if not isinstance(attributes, dict):
             raise TypeError("Dataset:add_rxn: attributes must be a dictionary, not '{}'".format(type(attributes)))
 
-        rxn["attributes"] = attributes
+        rxn_dict["attributes"] = attributes
 
         if not isinstance(other_fields, dict):
             raise TypeError("Dataset:add_rxn: other_fields must be a dictionary, not '{}'".format(type(attributes)))
 
         for k, v in other_fields.items():
-            rxn[k] = v
+            rxn_dict[k] = v
 
         if "default" in list(reaction_results):
-            rxn["reaction_results"] = reaction_results
+            rxn_dict["reaction_results"] = reaction_results
         elif isinstance(reaction_results, dict):
-            rxn["reaction_results"] = {}
-            rxn["reaction_results"]["default"] = reaction_results
+            rxn_dict["reaction_results"] = {}
+            rxn_dict["reaction_results"]["default"] = reaction_results
         else:
             raise TypeError("Passed in reaction_results not understood.")
 
-        self.data["reactions"].append(rxn)
+        rxn = Rxn(**rxn_dict)
+
+        self.data.reactions.append(rxn)
 
+        # TODO: Should this return the dict, or the PyDantic BaseModel?
```

Comment:
@dgasmith This is the one outstanding issue I have on this. The `self.data.reactions` is a list of `BaseModel` objects, not a list of `dict`. This leads to 2 questions:

1. Should this function, `add_rxn` return the `dict` or the `BaseModel`?
2. More broadly, this may change how users interact with the `self.data.rections`, which I imagine would eventually go to the more general Python `Dataclass` from 3.7 eventually anyways, which would behave mostly the same way. Should we expect our users to start doing this now, or should we continue to only expose the `dict` for `[key]` type access rather than by attribute?

Reply:
On a related note, I have the `get_rxn` function return a `dict`, not a `BaseModel`, which is inconsistent and we should pick one or the other. This is more a note for me but based on the answer above, will change what I do with it.

Reply:
1) Should return a BaseModel I think to be a bit more class focused.
2) Can you give an example? I would imagine users will want to do something like `self.data.reactions["my_cool_reaction"]`.

Will Dataclass eventually supersede pedantic?

Reply:
> Can you give an example? I would imagine users will want to do something like self.data.reactions["my_cool_reaction"].

Right, people would *want* to do that, but they cant. They can't do that right now to be fair.

The current (`master`, not this pr) model would be `self.data.reactions[3]['name']` to access the reaction name, the new, post-pr model would be `self.data.reactions[3]`.name` because `reactions` itself is still a list. This change may force users to alter their code a bit, but its a pretty trivial change, and one for the better I think. It does make arbitrary iteration a bit more odd since you would have to make this kind of change:

```python
my_atters = ['some', 'list', 'of', 'keys']

# old style
for reaction in data.reactions:
    result = [do_something(reaction[attr]) for atter in my_atters]

# new style
for reaction in data.reactions:
    result = [do_something(getattr(reaction, atter)) for atter in my_atters]
```
Although I imagine that being rare since I think the more common action will be:
```python
for reaction in data.reactions:
    result[reaction.name] = do_something(reaction.attr1, reaction.attr2)
```
which I think is less ambiguous and does not force the programmer/reader to internalize list comprehension and looping

> Will Dataclass eventually supersede pydantic?

No, because the `Dataclass` does not support validation and coercion, its just a container. The way I view the data class is the natural evolution the dict where:
* Keys are pure string,
* Keys are immutable
to a formal named class where the `key:value`  pair system is replaced with `Class.attribute = value`. If Python ever converted the `Dataclass` to enforce strict input types based on the type hints, then I could see pydantic loosing out eventually, but I don't think they will do that because then you may as well make variable declaration require type definition as well.

If this all makes sense to you, and you are okay with the users making this change, then I will make the following changes:
* Make `add_rxn` return the `Rxn(BaseModel)` class
* Make `get_rxn` return the `Rxn(BaseModel)` class

One small caveat is we don't protect the `Collection.data` object, or any of its values at all. There is nothing stopping users from modifying the exiting fields to whatever they want, but I don't think there is a strong need to do this.